src/jit/importer.cpp

```diff
@@ -3918,6 +3918,40 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_BitConverter_ByteSwap:
+            {
+                if (impStackTop(0).val->IsIntegralConst())
+                {
+                    // If the incoming value is a constant, don't bother with the intrinsic step.
+                    // Instead just let morph perform all of the bit shifting and calculate the final value.
+                    isSpecial = true;
+                    break;
+                }
+
+                // The BSWAP instruction is undefined for 16-bit inputs. Instead, use ROR 8 on the
+                // low word register bits. Additionally, BSWAP is undefined for 64-bit inputs on
```

Comment:
We could probably make decomposition generate 2 BSWAPs from a long BSWAP on x86.

Reply:
That's how this works internally. If the developer targets x86 but calls `swap(ulong)`, the call site remains a normal method call. The implementation of `swap(ulong)` is written in terms of `swap(uint)`, which gets correctly flagged as an intrinsic and converted to `bswap`.

The flip side is that this is the same reason `swap(<const ulong>)` doesn't currently get optimized properly, unfortunately. :(

Reply:
I see. I might make more sense to rely on JIT's decomposition on x86 though, if only because it avoids such ifdefs and inconsistencies between architectures. But it's not obvious what the best solution is, we'll see what the team thinks about it.

Reply:
I agree with @mikedn that it might be cleaner to handle the "longs on 32-bit target" in the JIT. However, I'm interested in what the perf issues are that you're seeing - is it because it interferes with inlining? Or that it generates extra copies? Or something else? The decomposition approach means that the front-end (importer and optimizations) of the JIT will treat these as operations on 64-bit values. Then, the backend "decomposes" these into the appropriate sequence of operations on 32-bit values. In general, this allows the JIT to do a better job of optimizing these in the front-end (though sometimes the decomposition itself can expose code sequences that might have been optimized if they had been expanded earlier; the classic "phase ordering" problem).

Reply:
The perf issue I mentioned was due to not enlightning morph's const folding logic to account for `GT_BSWAP`. This has been resolved with the latest iteration of the PR.

If you both still believe the best design is to have the JIT fully handle 64-bit byte swaps rather than falling back to the managed decomposition implementation, I can investigate that, but I need some assistance being pointed in the right direction. Is `CodeGen::genCodeForShiftLong` a good source of inspiration to turn to?